compiler/rustc_borrowck/src/type_check/mod.rs

```diff
@@ -427,14 +427,24 @@ impl<'a, 'b, 'tcx> Visitor<'tcx> for TypeVerifier<'a, 'b, 'tcx> {
                 // const_trait_impl: use a non-const param env when checking that a FnDef type is well formed.
                 // this is because the well-formedness of the function does not need to be proved to have `const`
                 // impls for trait bounds.
-                let instantiated_predicates = tcx.predicates_of(def_id).instantiate(tcx, substs);
                 let prev = self.cx.param_env;
                 self.cx.param_env = prev.without_const();
+
+                let instantiated_predicates = tcx.predicates_of(def_id).instantiate(tcx, substs);
                 self.cx.normalize_and_prove_instantiated_predicates(
                     def_id,
                     instantiated_predicates,
                     locations,
                 );
+
+                self.cx.prove_predicates(
```

Comment:
for this to be sound, `FnDef` during mir typeck always have to refer to the trait method, never to the impl method for trait methods.

E.g. in `src/test/ui/borrowck/fn-item-check-type-params.rs` the impl substs of `Displayable::display` in `test_call` is just `&'a str` which is WF.

I believe this to be the case, but that's an invariant we don't explicitly state anywhere iirc, so adding a debug assert for something like the following would be good:
```rust
if let Some(imp) = tcx.impl_of_method(def_id) {
    assert_eq!(tcx.impl_trait_ref(imp), None);
}
```

Reply:
Right! There is also a similar complication with inherent impls. Consider:
```rust
struct MyTy<T>(T);
impl<'a> MyTy<&'static &'a ()> {
    const CONST: fn(&'a str) -> &'static str = { |x| x };
    fn func(x: &'a str) -> &'static str { x } 
}
```
However this is not a problem currently because in order to refer to an associated function or const the user have to either explicitly annotate the `Self` type or, in the the case of method calls, the self type is included as an argument, and in both cases WF-check the `Self` type.